sky/provision/oci/query_utils.py

```diff
@@ -468,5 +470,97 @@ def delete_vcn(cls, net_client, skypilot_vcn, skypilot_subnet,
             logger.error(
                 f'Delete VCN {oci_utils.oci_config.VCN_NAME} Error: {str(e)}')
 
+    @classmethod
+    @debug_enabled(logger)
+    def add_ingress_rules(cls, region: str, ports: Set[int]):
+        if ports is None or len(ports) == 0:
+            return
+
+        net_client = oci_adaptor.get_net_client(
+            region, oci_utils.oci_config.get_profile())
+        subnet_id = query_helper.find_create_vcn_subnet(region)
+        get_subnet_resp = net_client.get_subnet(subnet_id=subnet_id)
+
+        existing_ports: Set[int] = set()
+        first_sg_id = None
+        first_sg_rules = None
+        security_list_ids = get_subnet_resp.data.security_list_ids
+
+        for sg_id in security_list_ids:
+            if first_sg_id is None:
+                first_sg_id = sg_id
+
+            get_sg_resp = net_client.get_security_list(security_list_id=sg_id)
+
+            ingress_rules: List = get_sg_resp.data.ingress_security_rules
+            if first_sg_rules is None:
+                first_sg_rules = ingress_rules
+
+            for r in ingress_rules:
+                rule_existing_ports = [
+                    p for p in ports if r.tcp_options and
+                    (r.tcp_options.destination_port_range.max >= p and
+                     r.tcp_options.destination_port_range.min <= p)
+                ]
+                existing_ports.update(rule_existing_ports)
+
+        new_ports = ports - existing_ports
+        new_rules = [
+            oci_adaptor.oci.core.models.IngressSecurityRule(
+                protocol='6',
+                source=oci_utils.oci_config.VCN_CIDR_INTERNET,
+                is_stateless=False,
+                source_type='CIDR_BLOCK',
+                tcp_options=oci_adaptor.oci.core.models.TcpOptions(
+                    destination_port_range=oci_adaptor.oci.core.models.
+                    PortRange(max=p, min=p),),
+                description=oci_utils.oci_config.SERVICE_PORT_RULE_TAG,
+            ) for p in new_ports
+        ]
+
+        if first_sg_id is not None and len(new_rules) > 0:
+            if first_sg_rules is None:
+                first_sg_rules = []
+
+            net_client.update_security_list(
+                security_list_id=first_sg_id,
+                update_security_list_details=oci_adaptor.oci.core.models.
+                UpdateSecurityListDetails(
+                    ingress_security_rules=(first_sg_rules + new_rules)),
+            )
```

Comment:
What will happen if the first sg id is None? Are we still gonna open the ports? IIUC it is possible when starting with a fresh environment where no rules exist.

Is it possible to simulate our AWS implementation, creating a security group for each cluster, that only applies to the cluster?

Reply:
Now I changed the implementation to open_ports/cleanup ports per cluster. And updated the PR comments to include "Test2"